java/monkey/src/main/java/root/parser/Parser.java

```diff
@@ -0,0 +1,112 @@
+package root.parser;
+
+import root.Token;
+import root.TokenType;
+import root.ast.*;
+import root.ast.expressions.IdentiferExpression;
+import root.ast.statements.LetStatement;
+import root.ast.statements.ReturnStatement;
+import root.ast.statements.Statement;
+import root.lexer.Lexer;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Parser {
+
+    private final Lexer lexer;
+
+    private Token currentToken;
+
+    private Token peekToken;
+
+    public final List<String> errors = new ArrayList<>();
+
+    public Parser(Lexer lexer) {
+        this.lexer = lexer;
+
+        this.nextToken();
+        this.nextToken();
+    }
+
+    public Program parseProgram() {
+        Program program = new Program();
+
+        while (!curTokenIs(TokenType.EOF)) {
+            try {
+                Statement statement = parseStatement();
+                if (statement != null) {
+                    program.getStatements().add(statement);
+                }
+            } catch (ParserException pe) {
```

Comment:
Hmm. Maybe an `ErrorStatement` might fit here in the future? Then we can do something like
```java
switch (statement) {
   case ErrorStatement es -> this.errors.add(statement.es.getMessage());
   case Statement s -> program.getStatements().add(statement); // Would make more sense to have an "addStatement(Statement)" function on Program
   default -> throw IllegalStateExecption("parseStatement() should never return a value to get here");
   nextToken();
}
```
That avoid the Parser Exception, and means if this throws an Exception, we're actually dealing with an Illegal State that the program isn't designed to handle.

Once again, we can do it in another PR.

Reply:
I don't know how I feel about an ErrorStatement, specially if it is a subclass of Statement, because statements in this contexts should be parts of the program the the interpreter will execute, an error caused by bad syntax is not a part of the program, it could never be added to the list of statements in the Program object or be executed, so I would strongly oppose it being a Statement, it should be it's own separate class. The bad part is that we would need to change the return types of the parsing functions to Object or something else, witch I also don't like, I honestly prefer just throwing checked exceptions in this case.